statsmodels/graphics/functional.py

```diff
@@ -1,14 +1,464 @@
+
+    Outliers are defined as curves that fall outside the band corresponding
+    to the quantile given by `threshold`.
+
+    The non-outlying region is defined as the band made up of all the
+    non-outlying curves.
+
+    Behind the scene, the dataset is represented as a matrix. Each line
+    corresponding to a 1D curve. This matrix is then decomposed using Principal
+    Components Analysis (PCA). This allows to represent the data using a finite
+    number of modes, or components. This compression process allows to turn the
+    functional representation into a scalar representation of the matrix. In
+    other words, you can visualize each curve from its components. Each curve
+    is thus a point in this reduced space. With 2 components, this is called a
+    bivariate plot (2D plot).
+
+    In this plot, if some points are adjacent (similar components), it means
+    that back in the original space, the curves are similar. Then, finding the
+    median curve means finding the higher density region (HDR) in the reduced
+    space. Moreover, the more you get away from this HDR, the more the curve is
+    unlikely to be similar to the other curves.
+
+    Using a kernel smoothing technique, the probability density function (PDF)
+    of the multivariate space can be recovered. From this PDF, it is possible to
+    compute the density probability linked to the cluster of points and plot
+    its contours.
+
+    Finally, using these contours, the different quantiles can be extracted
+    allong with the median curve and the outliers.
+
+    References
+    ----------
+    [1] R.J. Hyndman and H.L. Shang, "Rainbow Plots, Bagplots, and Boxplots for
+        Functional Data", vol. 19, pp. 29-45, 2010.
+
+    Examples
+    --------
+    Load the El Nino dataset.  Consists of 60 years worth of Pacific Ocean sea
+    surface temperature data.
+
+    >>> import matplotlib.pyplot as plt
+    >>> import statsmodels.api as sm
+    >>> data = sm.datasets.elnino.load()
+
+    Create a functional boxplot.  We see that the years 1982-83 and 1997-98 are
+    outliers; these are the years where El Nino (a climate pattern
+    characterized by warming up of the sea surface and higher air pressures)
+    occurred with unusual intensity.
+
+    >>> fig = plt.figure()
+    >>> ax = fig.add_subplot(111)
+    >>> res = sm.graphics.hdrboxplot(data.raw_data[:, 1:],
+    ...                              labels=data.raw_data[:, 0].astype(int),
+    ...                              ax=ax)
+
+    >>> ax.set_xlabel("Month of the year")
+    >>> ax.set_ylabel("Sea surface temperature (C)")
+    >>> ax.set_xticks(np.arange(13, step=3) - 1)
+    >>> ax.set_xticklabels(["", "Mar", "Jun", "Sep", "Dec"])
+    >>> ax.set_xlim([-0.2, 11.2])
+
+    >>> plt.show()
+
+    .. plot:: plots/graphics_functional_hdrboxplot.py
+
+    """
+    fig, ax = utils.create_mpl_ax(ax)
+
+    data = np.asarray(data)
+    if xdata is None:
+        xdata = np.arange(data.shape[1])
+
```

Comment:
suggestion for use with pandas, before converting to array on line 309, store the index as labels, e.g.
if isinstance(xdata, pd.DataFrame) and labels is None:
    labels = xdata.index
(maybe needs labels = np.asarray(xdata.index) )

alternative would be 
if hasattr(xdata, 'index') and labels is None:

Reply:
Done using your snippet bellow